### ISSUE TITLE
Android CI builds failing on release-ios-v3.4.0 branch

### ISSUE THREAD

**Issue by 1ec5** [mapbox/mapbox-gl-native#7763]
Android builds on Bitrise began failing on the release-ios-v3.4.0 branch (but not master), due to [a Bitrise VM update](https://discuss.bitrise.io/t/weekly-virtual-machine-stack-updates-2017-01-14/139), specifically bitrise-docker/android#55. For example, [this build](https://www.bitrise.io/build/6e0df94dbb89dd82) from #7759 failed with:

```
AWS Device Farm configuration is NOT VALID

FAILURE: Build failed with an exception.

* What went wrong:
A problem occurred configuring project ':MapboxGLAndroidSDKWearTestApp'.
> failed to find Build Tools revision 24.0.2
```

As a stopgap, since we’re very close to an iOS SDK release, I put in some triggers that diverted the Android build to a no-op workflow that always passes for branches pushed by @mapbox/ios. However, the triggers are based on branch name patterns, so for example I couldn’t find a good way to exempt @incanus from Android builds. Not that that would be a great idea: we’re starting to develop on master again, which means we’ll need Android builds to run; meanwhile, we still need to keep the release-ios-v3.4.0 branch alive for v3.4.1. We need to find a way to make these builds run for real but continue to pass.

/cc @mapbox/ios @mapbox/android

**Comment by friedbunny** [comment]
The issue here is that Bitrise stopped installing the version of build tools [that we specify](https://github.com/mapbox/mapbox-gl-native/blob/e38216d25444fd426921c372898c425d6030355e/platform/android/MapboxGLAndroidSDKWearTestApp/build.gradle#L24). There are two fixes that I see:

1. Add a commit that bumps the version number to one that Bitrise installs automatically, as in https://github.com/mapbox/mapbox-gl-native/pull/7729 on master. To avoid pulling in any potential change associated with 25.x, we could just bump to 24.0.3.
2. Install build tools 24.0.2 ourselves (via some Bitrise workflow step).

/cc @tobrun

**Comment by 1ec5** [comment]
Based on [this announcement](https://discuss.bitrise.io/t/android-automatic-dependency-installer-e-g-build-tools-step-install-missing-android-tools/316), I installed the “Install missing Android tools” step right after the “Git clone” step in the Android app on bitrise.io (not in bitrise.yml).

@tobrun, if this change checks out and you’d prefer to have the step in bitrise.yml, I can move it there, but I think this should be fine as a stopgap.

**Comment by 1ec5** [comment]
```
Failed to parse build.gradle at: /bitrise/src/platform/android/MapboxGLAndroidSDK/build.gradle, error: failed to determine compileSDKVersion, error: failed to parse compileSdkVersion (Integer.parseInt(project.ANDROID_BUILD_SDK_VERSION)), error: Malformed version: Integer.parseInt(project.ANDROID_BUILD_SDK_VERSION)
```

**Comment by 1ec5** [comment]
The step fails because [this line](https://github.com/mapbox/mapbox-gl-native/blob/release-ios-v3.4.0/platform/android/MapboxGLAndroidSDK/build.gradle#L38) refers to a constant in gradle.properties instead of the literal int that [the step is looking for](https://github.com/bitrise-steplib/steps-install-missing-android-tools/blob/master/analyzer/analyzer.go#L104).

**Comment by 1ec5** [comment]
Falling back to the third-party [Android SDK Update](https://github.com/guitcastro/steps-update-sdk/) step to install 24.0.2 explicitly.

**Comment by 1ec5** [comment]
Installing 24.0.2 explicitly did the trick. Note that all builds are affected, even builds on master that don’t use 24.0.2.

<img width="333" alt="install" src="https://cloud.githubusercontent.com/assets/1231218/22053124/8d08b190-dd03-11e6-8158-ac169cc6cbb8.png">

**Comment by tobrun** [comment]
I was aware of the capabilities shown above but I not apply this as it increases our build times and maintenance efforts. On the other hand it has the upside that commits in the git history will remain to work no matter what Bitrise does with future stack upgrades. I'm not opposed to this approach but have some small suggestions: 
 - read properties file containing the build tools version and use that as an input for the downloading  only that version of the tooling (removes maintenance efforts) 
 - make this step part of the versioned bitrise.yml

**Comment by zugaldia** [comment]
Why not simply cherrypicking https://github.com/mapbox/mapbox-gl-native/pull/7729 like @friedbunny suggests?

**Comment by 1ec5** [comment]
> read properties file containing the build tools version and use that as an input for the downloading only that version of the tooling (removes maintenance efforts)

That's what the Bitrise-provided step tried to do; however, it got tripped up trying to parse the SDK version as a literal int rather than an expression that referred to an entry in gradle.properties.

> Why not simply cherrypicking #7729 like @friedbunny suggests?

Totally open to this, as long as it applies cleanly to this branch. Reopening.

**Comment by friedbunny** [comment]
> Why not simply cherrypicking #7729 like @friedbunny suggests?

#7729 may require other changes, which could be a rabbit hole. It may work, but...

We should strive to keep historical commits (and current release branches, for that matter) passing on CI without additional commits. In this case, I think that means pursuing the second option I suggested: forcing the install of whichever dependencies we need and not relying on Bitrise to do so.

**Comment by 1ec5** [comment]
> We should strive to keep historical commits (and current release branches, for that matter) passing on CI without additional commits.

This is primarily important to facilitate bisecting. Fortunately, there’s no need to bisect Android issues on this branch, and commits on master have passed on Bitrise without interruption.

I very much liked the “Install missing Android tools” step that Bitrise provided, because it would protect future release branches against fly-by-night deployments on Bitrise’s end on a weekend. However, that step fails due to https://github.com/mapbox/mapbox-gl-native/issues/7763#issuecomment-273388275. @zugaldia, do you think it would be feasible for the step itself to be patched to parse non-literals? If so, I can file a ticket upstream.

**Comment by 1ec5** [comment]
Fixed in #7793 on the release-ios-v3.4.0 branch.

**Comment by 1ec5** [comment]
> I very much liked the “Install missing Android tools” step that Bitrise provided, because it would protect future release branches against fly-by-night deployments on Bitrise’s end on a weekend. However, that step fails due to https://github.com/mapbox/mapbox-gl-native/issues/7763#issuecomment-273388275.

The Bitrise team asked us to file bitrise-steplib/steps-install-missing-android-tools#5 about this issue.
